File: python/generator.py
"""Xorshift64 и генератор ордеров — побитово как golang/generator.go.

Работает вне таймера, скорость не важна; корректность u64-переполнения
обеспечивается маской (1 << 64) - 1.
"""

MASK64 = (1 << 64) - 1
DEFAULT_SEED = 0xDEADBEEFCAFE1234


def generate_orders(n, seed):
    """Возвращает четыре параллельных списка: sides, prices, qtys, ids.

    side: 0 = Buy, 1 = Sell; price: 9900..10099; qty: 1..100; id: i + 1.
    """
    v = seed & MASK64
    if v == 0:
        v = DEFAULT_SEED
    sides = [0] * n
    prices = [0] * n
    qtys = [0] * n
    ids = list(range(1, n + 1))
    m = MASK64
    for i in range(n):
        v ^= (v << 13) & m
        v ^= v >> 7
        v ^= (v << 17) & m
        r1 = v
        v ^= (v << 13) & m
        v ^= v >> 7
        v ^= (v << 17) & m
        r2 = v
        v ^= (v << 13) & m
        v ^= v >> 7
        v ^= (v << 17) & m
        r3 = v
        sides[i] = r1 & 1
        prices[i] = 10000 + (r2 % 200) - 100
        qtys[i] = 1 + (r3 % 100)
    return sides, prices, qtys, ids

File: nodejs/generator.js
'use strict';
// Xorshift64 на BigInt: побитово совпадает с golang/generator.go и rust/src/generator.rs.
// Работает ВНЕ таймера, поэтому скорость BigInt не важна; результат — плоские
// типизированные массивы (SoA), которые движок читает без объектов.

const MASK64 = (1n << 64n) - 1n;
const DEFAULT_SEED = 0xDEADBEEFCAFE1234n;

function generateOrders(n, seed) {
  let v = BigInt.asUintN(64, seed);
  if (v === 0n) v = DEFAULT_SEED;

  const side = new Uint8Array(n);
  const price = new Int32Array(n);
  const qty = new Int32Array(n);

  for (let i = 0; i < n; i++) {
    v ^= (v << 13n) & MASK64;
    v ^= v >> 7n;
    v ^= (v << 17n) & MASK64;
    const r1 = v;
    v ^= (v << 13n) & MASK64;
    v ^= v >> 7n;
    v ^= (v << 17n) & MASK64;
    const r2 = v;
    v ^= (v << 13n) & MASK64;
    v ^= v >> 7n;
    v ^= (v << 17n) & MASK64;
    const r3 = v;

    side[i] = Number(r1 & 1n);
    price[i] = 10000 + Number(r2 % 200n) - 100;
    qty[i] = 1 + Number(r3 % 100n);
  }
  return { n, side, price, qty };
}

module.exports = { generateOrders, DEFAULT_SEED };
